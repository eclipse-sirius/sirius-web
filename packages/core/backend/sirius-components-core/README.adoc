= sirius-components-core

== Goal

This project is used to contain the API to manipulate core concepts that are not specific to the business of our application.

== Dependencies

- sirius-components-annotations

This project should not depend on anything else without a good reason to do so.