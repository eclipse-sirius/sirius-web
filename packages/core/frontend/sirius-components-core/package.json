{
  "name": "@eclipse-sirius/sirius-components-core",
  "version": "2024.5.0",
  "author": "Eclipse Sirius",
  "license": "EPL-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/eclipse-sirius/sirius-web"
  },
  "publishConfig": {
    "registry": "https://npm.pkg.github.com/"
  },
  "main": "./dist/sirius-components-core.umd.js",
  "module": "./dist/sirius-components-core.es.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "exports": {
    ".": {
      "require": "./dist/sirius-components-core.umd.js",
      "import": "./dist/sirius-components-core.es.js"
    }
  },
  "scripts": {
    "start": "vite build --mode 'development' && tsc",
    "build": "vite build && tsc",
    "format": "prettier --write \"src/**/*.{js,ts,tsx,css}\"",
    "format-lint": "prettier --list-different \"src/**/*.{js,ts,tsx,css}\"",
    "test": "vitest --run --config vite.config.js",
    "coverage": "vitest run --config vite.config.js --coverage",
    "publish:local": "yalc push"
  },
  "peerDependencies": {
    "@apollo/client": "3.10.4",
    "@material-ui/core": "4.12.4",
    "@material-ui/icons": "4.11.3",
    "@xstate/react": "1.6.3",
    "graphql": "16.8.1",
    "react": "17.0.2",
    "xstate": "4.32.1"
  },
  "devDependencies": {
    "@apollo/client": "3.10.4",
    "@eclipse-sirius/sirius-components-tsconfig": "*",
    "@material-ui/core": "4.12.4",
    "@material-ui/icons": "4.11.3",
    "@types/react": "17.0.37",
    "@vitejs/plugin-react": "4.0.4",
    "@xstate/react": "1.6.3",
    "@vitest/coverage-v8": "0.34.2",
    "jsdom": "16.7.0",
    "graphql": "16.8.1",
    "prettier": "2.7.1",
    "react": "17.0.2",
    "rollup-plugin-peer-deps-external": "2.2.4",
    "xstate": "4.32.1",
    "typescript": "5.1.6",
    "vite": "4.4.9",
    "vitest": "0.34.2"
  }
}
