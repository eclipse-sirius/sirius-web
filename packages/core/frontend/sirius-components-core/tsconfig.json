{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "@eclipse-sirius/sirius-components-tsconfig/react-library.json",
  "include": ["."],
  "exclude": ["dist", "build", "node_modules"]
}
