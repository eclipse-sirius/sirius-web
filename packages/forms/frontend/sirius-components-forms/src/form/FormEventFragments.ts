/*******************************************************************************
 * Copyright (c) 2021, 2024 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/

export const commonFields = `
  fragment commonFields on Widget {
    id
    __typename
    diagnostics {
      id
      kind
      message
    }
    hasHelpText
    readOnly
  }
`;

export const widgetFields = `
  ${commonFields}
  
  fragment textfieldFields on Textfield {
    label
    iconURL
    stringValue: value
    supportsCompletion
    style {
      backgroundColor
      foregroundColor
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment textareaFields on Textarea {
    label
    iconURL
    stringValue: value
    supportsCompletion
    style {
      backgroundColor
      foregroundColor
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment checkboxFields on Checkbox {
    label
    iconURL
    booleanValue: value
    style {
      color
      labelPlacement
    }
  }

  fragment selectFields on Select {
    label
    iconURL
    value
    options {
      id
      label
      iconURL
    }
    style {
      backgroundColor
      foregroundColor
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment multiSelectFields on MultiSelect {
    label
    iconURL
    values
    options {
      id
      label
      iconURL
    }
    style {
      backgroundColor
      foregroundColor
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment radioFields on Radio {
    label
    iconURL
    options {
      id
      label
      selected
    }
    style {
      color
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment listFields on List {
    label
    iconURL
    items {
      id
      label
      kind
      iconURL
      deletable
    }
    style {
      color
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment linkFields on Link {
    label
    iconURL
    url
    style {
      color
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment buttonFields on Button {
    label
    buttonLabel
    imageURL
    style {
      backgroundColor
      foregroundColor
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment splitButtonFields on SplitButton {
    label
    actions {
      ...commonFields
      ...buttonFields
    }
  }

  fragment toolbarActionFields on ToolbarAction {
    label
    buttonLabel
    imageURL
    style {
      backgroundColor
      foregroundColor
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment chartWidgetFields on ChartWidget {
    label
    chart {
      __typename
      ... on BarChart {
        metadata {
          label
        }
        entries {
          key
          value
        }
        style {
          barsColor
          fontSize
          italic
          bold
          underline
          strikeThrough
        }
        width
        height
      }
      ... on PieChart {
        metadata {
          label
        }
        entries {
          key
          value
        }
        style {
          colors
          strokeColor
          strokeWidth
          fontSize
          italic
          bold
          underline
          strikeThrough
        }
      }
    }
  } 

  fragment labelWidgetFields on LabelWidget {
    label
    stringValue: value
    style {
      color
      fontSize
      italic
      bold
      underline
      strikeThrough
    }
  }

  fragment treeWidgetFields on TreeWidget {
    label
    iconURL
    expandedNodesIds
    nodes {
      id
      parentId
      label
      kind
      iconURL
      endIconsURL
      selectable
      checkable
      value
    }
  }

  fragment imageFields on Image {
    label
    iconURL
    url
    maxWidth
  }

  fragment richTextFields on RichText {
    label
    iconURL
    stringValue: value
  }

  fragment sliderFields on Slider {
    label
    iconURL
    minValue
    maxValue
    currentValue
  }
  
  fragment dateTimeFields on DateTime {
    label
    iconURL
    stringValue
    type
    style {
      backgroundColor
      foregroundColor
      italic
      bold
    }
  }

  fragment widgetFields on Widget {
    ...commonFields
    ... on Textfield {
      ...textfieldFields
    }
    ... on Textarea {
      ...textareaFields
    }
    ... on Checkbox {
      ...checkboxFields
    }
    ... on Select {
      ...selectFields
    }
    ... on MultiSelect {
      ...multiSelectFields
    }
    ... on Radio {
      ...radioFields
    }
    ... on List {
      ...listFields
    }
    ... on Link {
      ...linkFields
    }
    ... on Button {
      ...buttonFields
    }
    ... on SplitButton {
      ...splitButtonFields
    }
    ... on ToolbarAction {
      ...toolbarActionFields
    }
    ... on LabelWidget {
      ...labelWidgetFields
    }
    ... on ChartWidget {
      ...chartWidgetFields
    }
    ... on TreeWidget {
      ...treeWidgetFields
    }
    ... on Image {
      ...imageFields
    }
    ... on RichText {
      ...richTextFields
    }
    ... on Slider {
      ...sliderFields
    }
    ... on DateTime {
      ...dateTimeFields
    }
  }
`;

export const flexboxContainerFields = `
  ${commonFields}
  ${widgetFields}
  fragment flexboxContainerFields on FlexboxContainer {
    ...commonFields
    label
    flexDirection
    flexWrap
    flexGrow
    children {
      ...widgetFields
      ... on FlexboxContainer {
        ...commonFields
        label
        flexDirection
        flexWrap
        flexGrow
        children {
          ...widgetFields
          ... on FlexboxContainer {
            ...commonFields
            label
            flexDirection
            flexWrap
            flexGrow
            children {
              ...widgetFields
            }
            borderStyle {
              color
              lineStyle
              size
              radius
            }
          }
        }
        borderStyle {
          color
          lineStyle
          size
          radius
        }
      }
    }
    borderStyle {
      color
      lineStyle
      size
      radius
    }
  }
`;

export const formRefreshedEventPayloadFragment = `
  ${widgetFields}
  ${flexboxContainerFields}
  fragment formRefreshedEventPayloadFragment on FormRefreshedEventPayload {
    id
    form {
      id
      pages {
        id
        label
        groups {
          id
          label
          displayMode
          toolbarActions {
            ...commonFields
            ...toolbarActionFields
          }
          borderStyle {
            color
            lineStyle
            size
            radius
          }
          widgets {
            ...widgetFields
            ... on FlexboxContainer {
              ...flexboxContainerFields
            }
          }
        }
        toolbarActions {
            ...commonFields
            ...toolbarActionFields
        }
      }
    }
  }
`;
