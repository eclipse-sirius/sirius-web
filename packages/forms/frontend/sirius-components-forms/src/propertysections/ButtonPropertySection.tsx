/*******************************************************************************
 * Copyright (c) 2022, 2024 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { useMutation } from '@apollo/client';
import { ServerContext, ServerContextValue, getCSSColor, useMultiToast } from '@eclipse-sirius/sirius-components-core';
import Button from '@material-ui/core/Button';
import { Theme, makeStyles } from '@material-ui/core/styles';
import gql from 'graphql-tag';
import { useContext, useEffect } from 'react';
import { PropertySectionComponent, PropertySectionComponentProps } from '../form/Form.types';
import { GQLButton } from '../form/FormEventFragments.types';
import {
  ButtonStyleProps,
  GQLErrorPayload,
  GQLPushButtonInput,
  GQLPushButtonMutationData,
  GQLPushButtonMutationVariables,
  GQLPushButtonPayload,
  GQLSuccessPayload,
} from './ButtonPropertySection.types';
import { PropertySectionLabel } from './PropertySectionLabel';
import { getTextDecorationLineValue } from './getTextDecorationLineValue';

const useStyle = makeStyles<Theme, ButtonStyleProps>((theme) => ({
  style: {
    backgroundColor: ({ backgroundColor }) =>
      backgroundColor ? getCSSColor(backgroundColor, theme) : theme.palette.primary.light,
    color: ({ foregroundColor }) => (foregroundColor ? getCSSColor(foregroundColor, theme) : 'white'),
    fontSize: ({ fontSize }) => (fontSize ? fontSize : null),
    fontStyle: ({ italic }) => (italic ? 'italic' : null),
    fontWeight: ({ bold }) => (bold ? 'bold' : null),
    textDecorationLine: ({ underline, strikeThrough }) => getTextDecorationLineValue(underline, strikeThrough),
    paddingTop: theme.spacing(0.5),
    paddingBottom: theme.spacing(0.5),
    '&:hover': {
      backgroundColor: ({ backgroundColor }) =>
        backgroundColor ? getCSSColor(backgroundColor, theme) : theme.palette.primary.main,
      color: ({ foregroundColor }) => (foregroundColor ? getCSSColor(foregroundColor, theme) : 'white'),
      fontSize: ({ fontSize }) => (fontSize ? fontSize : null),
      fontStyle: ({ italic }) => (italic ? 'italic' : null),
      fontWeight: ({ bold }) => (bold ? 'bold' : null),
      textDecorationLine: ({ underline, strikeThrough }) => getTextDecorationLineValue(underline, strikeThrough),
    },
  },
  icon: {
    marginRight: ({ iconOnly }) => (iconOnly ? theme.spacing(0) : theme.spacing(2)),
  },
  propertySection: {
    display: 'flex',
    flexDirection: 'row',
    gap: theme.spacing(2),
  },
}));

export const pushButtonMutation = gql`
  mutation pushButton($input: PushButtonInput!) {
    pushButton(input: $input) {
      __typename
      ... on ErrorPayload {
        messages {
          body
          level
        }
      }
      ... on SuccessPayload {
        messages {
          body
          level
        }
      }
    }
  }
`;

const isErrorPayload = (payload: GQLPushButtonPayload): payload is GQLErrorPayload =>
  payload.__typename === 'ErrorPayload';

const isSuccessPayload = (payload: GQLPushButtonPayload): payload is GQLSuccessPayload =>
  payload.__typename === 'SuccessPayload';

/**
 * Defines the content of a Button property section.
 * The content is submitted when pushing the button.
 */
export const ButtonPropertySection: PropertySectionComponent<GQLButton> = ({
  editingContextId,
  formId,
  widget,
  readOnly,
}: PropertySectionComponentProps<GQLButton>) => {
  const props: ButtonStyleProps = {
    backgroundColor: widget.style?.backgroundColor ?? null,
    foregroundColor: widget.style?.foregroundColor ?? null,
    fontSize: widget.style?.fontSize ?? null,
    italic: widget.style?.italic ?? null,
    bold: widget.style?.bold ?? null,
    underline: widget.style?.underline ?? null,
    strikeThrough: widget.style?.strikeThrough ?? null,
    iconOnly: widget.buttonLabel ? false : true,
  };
  const classes = useStyle(props);

  const { httpOrigin } = useContext<ServerContextValue>(ServerContext);

  const [pushButton, { loading, data, error }] = useMutation<GQLPushButtonMutationData, GQLPushButtonMutationVariables>(
    pushButtonMutation
  );

  const { addErrorMessage, addMessages } = useMultiToast();

  useEffect(() => {
    if (!loading) {
      if (error) {
        addErrorMessage('An unexpected error has occurred, please refresh the page');
      }
      if (data) {
        const { pushButton } = data;
        if (isErrorPayload(pushButton) || isSuccessPayload(pushButton)) {
          addMessages(pushButton.messages);
        }
      }
    }
  }, [loading, error, data]);

  const onClick = () => {
    const input: GQLPushButtonInput = {
      id: crypto.randomUUID(),
      editingContextId,
      representationId: formId,
      buttonId: widget.id,
    };
    const variables: GQLPushButtonMutationVariables = { input };
    pushButton({ variables });
  };

  return (
    <div className={classes.propertySection}>
      <PropertySectionLabel
        editingContextId={editingContextId}
        formId={formId}
        widget={widget}
        data-testid={widget.label}
      />
      <Button
        data-testid={widget.buttonLabel}
        variant="contained"
        color="primary"
        onClick={onClick}
        disabled={readOnly || widget.readOnly}
        classes={{ root: classes.style }}>
        {widget.imageURL?.length > 0 ? (
          <img className={classes.icon} width="16" height="16" alt={widget.label} src={httpOrigin + widget.imageURL} />
        ) : null}
        {widget.buttonLabel}
      </Button>
    </div>
  );
};
