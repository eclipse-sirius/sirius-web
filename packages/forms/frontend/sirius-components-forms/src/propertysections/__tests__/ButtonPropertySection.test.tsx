/*******************************************************************************
 * Copyright (c) 2022, 2023 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { ServerContext } from '@eclipse-sirius/sirius-components-core';
import { act, cleanup, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { afterEach, expect, test, vi } from 'vitest';
import { GQLButton } from '../../form/FormEventFragments.types';
import { ButtonPropertySection, pushButtonMutation, updateWidgetFocusMutation } from '../ButtonPropertySection';
import {
  GQLErrorPayload,
  GQLPushButtonMutationData,
  GQLPushButtonMutationVariables,
  GQLSuccessPayload,
  GQLUpdateWidgetFocusMutationData,
  GQLUpdateWidgetFocusMutationVariables,
  GQLUpdateWidgetFocusSuccessPayload,
} from '../ButtonPropertySection.types';

vi.mock('uuid', () => ({ v4: () => '48be95fc-3422-45d3-b1f9-d590e847e9e1' }));

afterEach(() => cleanup());

const defaultButton: GQLButton = {
  __typename: 'Button',
  id: 'buttonId',
  label: 'Label',
  iconURL: null,
  diagnostics: [],
  buttonLabel: 'ButtonLabel',
  imageURL: null,
  style: null,
};

const buttonWithStyle: GQLButton = {
  __typename: 'Button',
  id: 'buttonId',
  label: 'Label',
  iconURL: null,
  diagnostics: [],
  buttonLabel: 'ButtonLabel',
  imageURL: null,
  style: {
    backgroundColor: '#de1000',
    foregroundColor: '#fbb800',
    fontSize: 14,
    italic: false,
    bold: false,
    underline: false,
    strikeThrough: false,
  },
};

const buttonWithEmptyStyle: GQLButton = {
  __typename: 'Button',
  id: 'buttonId',
  label: 'Label',
  iconURL: null,
  diagnostics: [],
  buttonLabel: 'ButtonLabel',
  imageURL: null,
  style: {
    backgroundColor: '',
    foregroundColor: '',
    fontSize: 14,
    italic: false,
    bold: false,
    underline: false,
    strikeThrough: false,
  },
};

const pushButtonVariables: GQLPushButtonMutationVariables = {
  input: {
    id: '48be95fc-3422-45d3-b1f9-d590e847e9e1',
    editingContextId: 'editingContextId',
    representationId: 'formId',
    buttonId: 'buttonId',
  },
};
const successPayload: GQLSuccessPayload = { __typename: 'SuccessPayload' };
const pushButtonSuccessData: GQLPushButtonMutationData = { pushButton: successPayload };

const pushButtonErrorPayload: GQLErrorPayload = {
  __typename: 'ErrorPayload',
  message: 'An error has occurred, please refresh the page',
};
const pushButtonErrorData: GQLPushButtonMutationData = {
  pushButton: pushButtonErrorPayload,
};

const updateWidgetFocusVariables: GQLUpdateWidgetFocusMutationVariables = {
  input: {
    id: '48be95fc-3422-45d3-b1f9-d590e847e9e1',
    editingContextId: 'editingContextId',
    representationId: 'formId',
    widgetId: 'buttonId',
    selected: true,
  },
};
const updateWidgetFocusSuccessPayload: GQLUpdateWidgetFocusSuccessPayload = {
  __typename: 'UpdateWidgetFocusSuccessPayload',
};
const updateWidgetFocusSuccessData: GQLUpdateWidgetFocusMutationData = {
  updateWidgetFocus: updateWidgetFocusSuccessPayload,
};

const updateWidgetFocusErrorPayload: GQLErrorPayload = {
  __typename: 'ErrorPayload',
  message: 'An error has occurred, please refresh the page',
};
const updateWidgetFocusErrorData: GQLUpdateWidgetFocusMutationData = {
  updateWidgetFocus: updateWidgetFocusErrorPayload,
};

test('should render the button', () => {
  const { container } = render(
    <MockedProvider>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ButtonPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={defaultButton}
          subscribers={[]}
          readOnly={false}
        />
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();
});

test('should render a readOnly button', () => {
  const { container } = render(
    <MockedProvider>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ButtonPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={defaultButton}
          subscribers={[]}
          readOnly
        />
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();
});

test('should send mutation when clicked', async () => {
  let updateWidgetFocusCalled = false;
  const updateWidgetFocusSuccessMock: MockedResponse<Record<string, any>> = {
    request: {
      query: updateWidgetFocusMutation,
      variables: updateWidgetFocusVariables,
    },
    result: () => {
      updateWidgetFocusCalled = true;
      return { data: updateWidgetFocusSuccessData };
    },
  };

  let pushButtonCalled = false;
  const pushButtonSuccessMock: MockedResponse<Record<string, any>> = {
    request: {
      query: pushButtonMutation,
      variables: pushButtonVariables,
    },
    result: () => {
      pushButtonCalled = true;
      return { data: pushButtonSuccessData };
    },
  };

  const mocks = [updateWidgetFocusSuccessMock, pushButtonSuccessMock];
  const { container } = render(
    <MockedProvider mocks={mocks}>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ButtonPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={defaultButton}
          subscribers={[]}
          readOnly={false}
        />
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();

  const element: HTMLElement = screen.getByTestId('ButtonLabel');

  await act(async () => {
    userEvent.click(element);
    await new Promise((resolve) => setTimeout(resolve, 0));

    await waitFor(() => {
      expect(updateWidgetFocusCalled).toBeTruthy();
      expect(pushButtonCalled).toBeTruthy();
      expect(container).toMatchSnapshot();
    });
  });
});

test('should display the error received', async () => {
  let updateWidgetFocusCalled = false;
  const updateWidgetFocusErrorMock: MockedResponse<Record<string, any>> = {
    request: {
      query: updateWidgetFocusMutation,
      variables: updateWidgetFocusVariables,
    },
    result: () => {
      updateWidgetFocusCalled = true;
      return { data: updateWidgetFocusErrorData };
    },
  };

  let pushButtonCalled = false;
  const pushButtonErrorMock: MockedResponse<Record<string, any>> = {
    request: {
      query: pushButtonMutation,
      variables: pushButtonVariables,
    },
    result: () => {
      pushButtonCalled = true;
      return { data: pushButtonErrorData };
    },
  };

  const mocks = [updateWidgetFocusErrorMock, pushButtonErrorMock];
  const { baseElement } = render(
    <MockedProvider mocks={mocks}>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ButtonPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={defaultButton}
          subscribers={[]}
          readOnly={false}
        />
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();

  const element: HTMLElement = screen.getByTestId('ButtonLabel');

  await act(async () => {
    userEvent.click(element);
    await new Promise((resolve) => setTimeout(resolve, 0));

    await waitFor(() => {
      expect(updateWidgetFocusCalled).toBeTruthy();
      expect(pushButtonCalled).toBeTruthy();
      expect(screen.getByTestId('error')).not.toBeUndefined();
      expect(baseElement).toMatchSnapshot();
    });
  });
});

test('should render the button without style', () => {
  const { baseElement } = render(
    <MockedProvider>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ButtonPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={defaultButton}
          subscribers={[]}
          readOnly={false}
        />
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});

test('should render the button with style', () => {
  const { baseElement } = render(
    <MockedProvider>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ButtonPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={buttonWithStyle}
          subscribers={[]}
          readOnly={false}
        />
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});

test('should render the button with empty style', async () => {
  const { baseElement } = render(
    <MockedProvider>
      <ServerContext.Provider value={{ httpOrigin: 'http://localhost' }}>
        <ButtonPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={buttonWithEmptyStyle}
          subscribers={[]}
          readOnly={false}
        />
      </ServerContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});
