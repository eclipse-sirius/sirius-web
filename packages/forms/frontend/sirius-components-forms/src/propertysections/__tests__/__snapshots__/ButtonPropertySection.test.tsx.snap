// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should display the error received 1`] = `
<body>
  <div>
    <div
      class="makeStyles-propertySection-30"
    >
      <div
        class="makeStyles-propertySectionLabel-33"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-36 MuiTypography-subtitle2"
        >
          Label
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-28 makeStyles-style-31 MuiButton-contained MuiButton-containedPrimary"
        data-testid="ButtonLabel"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`should display the error received 2`] = `
<body>
  <div>
    <div
      class="makeStyles-propertySection-30"
    >
      <div
        class="makeStyles-propertySectionLabel-33"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-36 MuiTypography-subtitle2"
        >
          Label
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-28 makeStyles-style-31 MuiButton-contained MuiButton-containedPrimary"
        data-testid="ButtonLabel"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`should render a readOnly button 1`] = `
<div>
  <div
    class="makeStyles-propertySection-12"
  >
    <div
      class="makeStyles-propertySectionLabel-15"
    >
      <h6
        class="MuiTypography-root makeStyles-typography-18 MuiTypography-subtitle2"
      >
        Label
      </h6>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root makeStyles-style-10 makeStyles-style-13 MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
      data-testid="ButtonLabel"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        ButtonLabel
      </span>
    </button>
  </div>
</div>
`;

exports[`should render a readOnly button from widget properties 1`] = `
<body>
  <div>
    <div
      class="makeStyles-propertySection-75"
    >
      <div
        class="makeStyles-propertySectionLabel-78"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-81 MuiTypography-subtitle2"
        >
          Label
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-73 makeStyles-style-76 MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        data-testid="ButtonLabel"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
      </button>
    </div>
  </div>
</body>
`;

exports[`should render the button 1`] = `
<div>
  <div
    class="makeStyles-propertySection-3"
  >
    <div
      class="makeStyles-propertySectionLabel-6"
    >
      <h6
        class="MuiTypography-root makeStyles-typography-9 MuiTypography-subtitle2"
      >
        Label
      </h6>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root makeStyles-style-1 makeStyles-style-4 MuiButton-contained MuiButton-containedPrimary"
      data-testid="ButtonLabel"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        ButtonLabel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`should render the button with empty style 1`] = `
<body>
  <div>
    <div
      class="makeStyles-propertySection-57"
    >
      <div
        class="makeStyles-propertySectionLabel-60"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-63 MuiTypography-subtitle2"
        >
          Label
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-55 makeStyles-style-58 MuiButton-contained MuiButton-containedPrimary"
        data-testid="ButtonLabel"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`should render the button with help hint 1`] = `
<body>
  <div>
     
    <div
      class="makeStyles-propertySection-66"
    >
      <div
        class="makeStyles-propertySectionLabel-69"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-72 MuiTypography-subtitle2"
        >
          Label
        </h6>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
          focusable="false"
          style="margin-left: 8px; font-size: 16px;"
          title=""
          viewBox="0 0 24 24"
        >
          <path
            d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
          />
        </svg>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-64 makeStyles-style-67 MuiButton-contained MuiButton-containedPrimary"
        data-testid="ButtonLabel"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`should render the button with style 1`] = `
<body>
  <div>
    <div
      class="makeStyles-propertySection-48"
    >
      <div
        class="makeStyles-propertySectionLabel-51"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-54 MuiTypography-subtitle2"
        >
          Label
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-46 makeStyles-style-49 MuiButton-contained MuiButton-containedPrimary"
        data-testid="ButtonLabel"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`should render the button without style 1`] = `
<body>
  <div>
    <div
      class="makeStyles-propertySection-39"
    >
      <div
        class="makeStyles-propertySectionLabel-42"
      >
        <h6
          class="MuiTypography-root makeStyles-typography-45 MuiTypography-subtitle2"
        >
          Label
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-style-37 makeStyles-style-40 MuiButton-contained MuiButton-containedPrimary"
        data-testid="ButtonLabel"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ButtonLabel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`should send mutation when clicked 1`] = `
<div>
  <div
    class="makeStyles-propertySection-21"
  >
    <div
      class="makeStyles-propertySectionLabel-24"
    >
      <h6
        class="MuiTypography-root makeStyles-typography-27 MuiTypography-subtitle2"
      >
        Label
      </h6>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root makeStyles-style-19 makeStyles-style-22 MuiButton-contained MuiButton-containedPrimary"
      data-testid="ButtonLabel"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        ButtonLabel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`should send mutation when clicked 2`] = `
<div>
  <div
    class="makeStyles-propertySection-21"
  >
    <div
      class="makeStyles-propertySectionLabel-24"
    >
      <h6
        class="MuiTypography-root makeStyles-typography-27 MuiTypography-subtitle2"
      >
        Label
      </h6>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root makeStyles-style-19 makeStyles-style-22 MuiButton-contained MuiButton-containedPrimary"
      data-testid="ButtonLabel"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        ButtonLabel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
